nasa-pictures-stealer/src/main/java/nasapicturesstealer/infrastructure/repository/NasaPicturesRepository.java

```diff
@@ -0,0 +1,72 @@
+package nasapicturesstealer.infrastructure.repository;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import lombok.RequiredArgsConstructor;
+import nasapicturesstealer.domain.entity.NasaCamera;
+import nasapicturesstealer.domain.entity.NasaPicture;
+import nasapicturesstealer.infrastructure.entity.NasaCameraEntity;
+import nasapicturesstealer.infrastructure.entity.NasaPictureEntity;
+import nasapicturesstealer.infrastructure.mapper.NasaCameraMapper;
+import nasapicturesstealer.infrastructure.mapper.NasaPictureMapper;
+import org.springframework.data.util.Pair;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class NasaPicturesRepository {
+
+    private final NasaPictureEntityRepository nasaPictureEntityRepository;
+  
+    private final NasaCameraEntityRepository nasaCameraEntityRepository;
+    
+    private final NasaPictureMapper nasaPictureMapper;
+    
+    private final NasaCameraMapper nasaCameraMapper;
+    
+    public List<NasaPicture> saveDomainEntities(List<NasaPicture> pictures) {
+      Map<Integer, Long> nasaIdToPictureId = this.nasaPictureEntityRepository
+        .findAllByNasaIds(pictures.stream().map(NasaPicture::getNasaId).toList())
+        .stream()
+        .collect(Collectors.toMap(Pair::getFirst, Pair::getSecond));
+
+      Map<Integer, Long> nasaIdToCameraId = this.nasaCameraEntityRepository
```

Comment:
This method is overloaded with logic that solves nothing.  It's very hard to read and I've been not even sure this code was working before start it. 
 
All this code can be replaced with a more transparent one, plus you need to add caching and transactionality

```suggestion
     public List<NasaPicture> saveDomainEntities(List<NasaPicture> pictures) {
        Map<NasaCamera, List<NasaPicture>> cameraToPicturesMap =
                pictures.stream().collect(Collectors.groupingBy(NasaPicture::getNasaCamera));

        return cameraToPicturesMap.entrySet().stream()
                                  .flatMap(entry -> processCameraPictures(entry).stream())
                                  .collect(Collectors.toList());
    }

    private List<NasaPicture> processCameraPictures(Map.Entry<NasaCamera, List<NasaPicture>> entry) {
        final NasaCamera camera = entry.getKey();
        final List<NasaPicture> nasaPictures = entry.getValue();

        NasaCameraEntity nasaCameraEntity = findOrCreateCameraEntity(camera);

        return nasaPictures.stream()
                           .map(picture -> findOrCreatePictureEntity(picture,
                                                                     nasaCameraEntity))
                           .map(nasaPictureMapper::toDomainEntity)
                           .collect(Collectors.toList());
    }

    private NasaCameraEntity findOrCreateCameraEntity(NasaCamera camera) {
        return nasaCameraEntityRepository.findByNasaId(camera.getNasaId())
                                         .orElseGet(() -> nasaCameraEntityRepository.save(nasaCameraMapper.toEntity(camera)));
    }

    private NasaPictureEntity findOrCreatePictureEntity(NasaPicture picture,
                                                        NasaCameraEntity cameraEntity) {
        return nasaPictureEntityRepository.findByNasaId(picture.getNasaId())
                                          .orElseGet(() -> {
                                              NasaPictureEntity entity = new NasaPictureEntity(cameraEntity,
                                                                                               picture.getNasaId(),
                                                                                               picture.getImgSrc());
                                              return nasaPictureEntityRepository.save(entity);
                                          });
    }
```